File: bigbluebutton-apps/src/main/scala/org/bigbluebutton/core/apps/chat/red5/ChatClientMessageSender.scala
package org.bigbluebutton.core.apps.chat.red5

import org.bigbluebutton.conference.meeting.messaging.red5.ConnectionInvokerService
import org.bigbluebutton.core.api._
import org.bigbluebutton.conference.meeting.messaging.red5.DirectClientMessage
import com.google.gson.Gson
import org.bigbluebutton.conference.meeting.messaging.red5.BroadcastClientMessage
import scala.collection.mutable.HashMap
import collection.JavaConverters._
import scala.collection.JavaConversions._
import java.util.ArrayList

class ChatClientMessageSender(service: ConnectionInvokerService) extends OutMessageListener2 {
 
	def handleMessage(msg: IOutMessage) {
	  msg match {
	    case msg: GetChatHistoryReply               => handleGetChatHistoryReply(msg)
	    case msg: SendPublicMessageEvent            => handleSendPublicMessageEvent(msg)
	    case msg: SendPrivateMessageEvent           => handleSendPrivateMessageEvent(msg)
	    case _ => // do nothing
	  }
	}   
  
  private def handleGetChatHistoryReply(msg: GetChatHistoryReply) {	
    val gson = new Gson();
    val message = new java.util.HashMap[String, Object]()
  	
  	val collection = new ArrayList[java.util.Map[String, String]]();
  	  
  	msg.history.foreach(p => {
  	    collection.add(mapAsJavaMap(p))
  	})
  	
  	val jsonMsg = gson.toJson(collection)
  	
//  	System.out.println("************ CHAT HISTORY = \n" + jsonMsg + "\n")
	  message.put("msg", jsonMsg)
  	  
	  val m = new DirectClientMessage(msg.meetingID, msg.requesterID, "ChatRequestMessageHistoryReply", message);
	  service.sendMessage(m);
  }
  
  private def handleSendPublicMessageEvent(msg: SendPublicMessageEvent) { 
    val gson = new Gson()
	  val jsonMsg = gson.toJson(mapAsJavaMap(msg.message))
//	System.out.println("************ PUBLIC CHAT MESSAGE = \n" + jsonMsg + "\n")
	  
	val m = new BroadcastClientMessage(msg.meetingID, "ChatReceivePublicMessageCommand", mapAsJavaMap(msg.message));
	service.sendMessage(m);    
  }
  
  private def handleSendPrivateMessageEvent(msg: SendPrivateMessageEvent) {
    msg.message.get("toUserID") match {
      case Some(uid) => {
 		val m = new DirectClientMessage(msg.meetingID, uid, "ChatReceivePrivateMessageCommand", mapAsJavaMap(msg.message));
		service.sendMessage(m);         
      }
      case None => // do nothing
    
    }
    	
	val m2 = new DirectClientMessage(msg.meetingID, msg.requesterID, "ChatReceivePrivateMessageCommand", mapAsJavaMap(msg.message));
	service.sendMessage(m2);    
  }

}

File: bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/client/ChatClientMessageSender.java
package org.bigbluebutton.red5.client;

import java.util.HashMap;
import java.util.Map;

import org.bigbluebutton.common.messages.GetChatHistoryReplyMessage;
import org.bigbluebutton.common.messages.SendPrivateChatMessage;
import org.bigbluebutton.common.messages.SendPublicChatMessage;
import org.bigbluebutton.red5.client.messaging.BroadcastClientMessage;
import org.bigbluebutton.red5.client.messaging.ConnectionInvokerService;
import org.bigbluebutton.red5.client.messaging.DirectClientMessage;
import org.bigbluebutton.red5.service.ChatKeyUtil;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

public class ChatClientMessageSender {
	private ConnectionInvokerService service;
	
	public ChatClientMessageSender(ConnectionInvokerService service) {
		this.service = service;
	}

	public void handleChatMessage(String message) {

		JsonParser parser = new JsonParser();
		JsonObject obj = (JsonObject) parser.parse(message);

		if (obj.has("header") && obj.has("payload")) {
			JsonObject header = (JsonObject) obj.get("header");

			if (header.has("name")) {
				String messageName = header.get("name").getAsString();
				switch (messageName) {
					case SendPublicChatMessage.SEND_PUBLIC_CHAT_MESSAGE:
						SendPublicChatMessage spucm = SendPublicChatMessage.fromJson(message);

						if (spucm != null) {
							processSendPublicChatMessage(spucm);
						}
						break;
					case SendPrivateChatMessage.SEND_PRIVATE_CHAT_MESSAGE:
						SendPrivateChatMessage sprcm = SendPrivateChatMessage.fromJson(message);

						if (sprcm != null) {
							processSendPrivateChatMessage(sprcm);
						}
						break;
					case GetChatHistoryReplyMessage.GET_CHAT_HISTORY_REPLY:
						GetChatHistoryReplyMessage gch = GetChatHistoryReplyMessage.fromJson(message);

						if (gch != null) {
							processGetChatHistoryReply(gch);
						}
						break;
				}
			}
		}
	}


	private void processSendPublicChatMessage(SendPublicChatMessage msg) {
		Map<String, Object> messageInfo = new HashMap<String, Object>();
		messageInfo.put(ChatKeyUtil.CHAT_TYPE, msg.messageInfo.get(ChatKeyUtil.CHAT_TYPE));
		messageInfo.put(ChatKeyUtil.FROM_USERID, msg.messageInfo.get(ChatKeyUtil.FROM_USERID));
		messageInfo.put(ChatKeyUtil.FROM_USERNAME, msg.messageInfo.get(ChatKeyUtil.FROM_USERNAME));
		messageInfo.put(ChatKeyUtil.TO_USERID, msg.messageInfo.get(ChatKeyUtil.TO_USERID));
		messageInfo.put(ChatKeyUtil.TO_USERNAME, msg.messageInfo.get(ChatKeyUtil.TO_USERNAME));
		messageInfo.put(ChatKeyUtil.FROM_TIME, msg.messageInfo.get(ChatKeyUtil.FROM_TIME));
		messageInfo.put(ChatKeyUtil.FROM_TZ_OFFSET, msg.messageInfo.get(ChatKeyUtil.FROM_TZ_OFFSET));
		messageInfo.put(ChatKeyUtil.FROM_COLOR, msg.messageInfo.get(ChatKeyUtil.FROM_COLOR));
		messageInfo.put(ChatKeyUtil.MESSAGE, msg.messageInfo.get(ChatKeyUtil.MESSAGE));

		BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "ChatReceivePublicMessageCommand", messageInfo);
		service.sendMessage(m);
	}

	private void processSendPrivateChatMessage(SendPrivateChatMessage msg) {
		Map<String, Object> messageInfo = new HashMap<String, Object>();
		messageInfo.put(ChatKeyUtil.CHAT_TYPE, msg.messageInfo.get(ChatKeyUtil.CHAT_TYPE));
		messageInfo.put(ChatKeyUtil.FROM_USERID, msg.messageInfo.get(ChatKeyUtil.FROM_USERID));
		messageInfo.put(ChatKeyUtil.FROM_USERNAME, msg.messageInfo.get(ChatKeyUtil.FROM_USERNAME));
		messageInfo.put(ChatKeyUtil.TO_USERID, msg.messageInfo.get(ChatKeyUtil.TO_USERID));
		messageInfo.put(ChatKeyUtil.TO_USERNAME, msg.messageInfo.get(ChatKeyUtil.TO_USERNAME));
		messageInfo.put(ChatKeyUtil.FROM_TIME, msg.messageInfo.get(ChatKeyUtil.FROM_TIME));
		messageInfo.put(ChatKeyUtil.FROM_TZ_OFFSET, msg.messageInfo.get(ChatKeyUtil.FROM_TZ_OFFSET));
		messageInfo.put(ChatKeyUtil.FROM_COLOR, msg.messageInfo.get(ChatKeyUtil.FROM_COLOR));
		messageInfo.put(ChatKeyUtil.MESSAGE, msg.messageInfo.get(ChatKeyUtil.MESSAGE));

		String toUserId = msg.messageInfo.get(ChatKeyUtil.TO_USERID);
		DirectClientMessage receiver = new DirectClientMessage(msg.meetingId, toUserId, "ChatReceivePrivateMessageCommand", messageInfo);
		service.sendMessage(receiver);
		
		DirectClientMessage sender = new DirectClientMessage(msg.meetingId, msg.requesterId, "ChatReceivePrivateMessageCommand", messageInfo);
		service.sendMessage(sender);
	}

	private void processGetChatHistoryReply(GetChatHistoryReplyMessage gch) {

		Map<String, Object> args = new HashMap<String, Object>();	
		args.put("meetingId", gch.meetingId);
		args.put("requester_id", gch.requesterId);
		args.put("chat_history", gch.chatHistory);

		Map<String, Object> message = new HashMap<String, Object>();
		Gson gson = new Gson();
		message.put("msg", gson.toJson(args.get("chat_history")));

		DirectClientMessage m = new DirectClientMessage(gch.meetingId, gch.requesterId, "ChatRequestMessageHistoryReply", message);
		service.sendMessage(m);
	}

}
